---
layout:         post
title:          DotNet
category:       blog
description:    总结一些DotNet编程中的trick
---

## Trick

### Try-Finally
- Case 295905415080611
	- Symptom: .NET application hang sometimes
	- Analysis:
		- Capture hang dump
		- Callstack

				0:000> kL
				# Child-SP          RetAddr           Call Site
				00 00000000`002ac0d8 000007fe`fd8e1430 ntdll!ZwWaitForMultipleObjects+0xa
				01 00000000`002ac0e0 00000000`776906b0 KERNELBASE!WaitForMultipleObjectsEx+0xe8
				02 00000000`002ac1e0 000007fe`f8d389a9 kernel32!WaitForMultipleObjects+0xb0
				03 00000000`002ac270 000007fe`f8d892d2 mscorwks!Debugger::EnsureDebuggerAttached+0xe9
				04 00000000`002ac300 000007fe`f8ef08d0 mscorwks!Debugger::SendException+0x1e2
				05 00000000`002ac8d0 000007fe`f8c66190 mscorwks!Debugger::LastChanceManagedException+0x1c0
				06 00000000`002ac9d0 000007fe`f8a0d154 mscorwks!NotifyDebuggerLastChance+0x60
				07 00000000`002aca20 000007fe`f8e6dadb mscorwks!WatsonLastChance+0x1a4
				08 00000000`002aca90 000007fe`f8e6fac9 mscorwks!InternalUnhandledExceptionFilter_Worker+0x22b
				09 00000000`002acb30 00000000`7771bc00 mscorwks!InternalUnhandledExceptionFilter+0x29
				0a 00000000`002acb60 00000000`7791ff88 kernel32!UnhandledExceptionFilter+0x160
				0b 00000000`002acc40 00000000`778b78c8 ntdll!RtlUserThreadStart$filt$0+0x45
				0c 00000000`002acc70 00000000`778c7e8d ntdll!__C_specific_handler+0x8c
				0d 00000000`002acce0 00000000`778b84cf ntdll!RtlpExecuteHandlerForException+0xd
				0e 00000000`002acd10 00000000`778b8ac8 ntdll!RtlDispatchException+0x45a
				0f 00000000`002ad3f0 000007fe`fd8fa06d ntdll!RtlRaiseException+0x22f
				10 00000000`002adda0 000007fe`f88bb943 KERNELBASE!RaiseException+0x39
				11 00000000`002ade70 000007fe`f8e61cf0 mscorwks!RaiseTheExceptionInternalOnly+0x2ff
				12 00000000`002adf60 000007fe`f858bada mscorwks!JIT_Throw+0x130
				13 00000000`002ae110 000007fe`f7bdac5b mscorlib_ni!System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(System.Runtime.Remoting.Messaging.IMessage, System.Runtime.Remoting.Messaging.IMessage)+0x9ca75a
				14 00000000`002ae150 000007fe`f89ccbd5 mscorlib_ni!System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(System.Runtime.Remoting.Proxies.MessageData ByRef, Int32)+0x21b
				15 00000000`002ae360 000007fe`f8c3b6eb mscorwks!CallTargetWorker2+0xc1
				16 00000000`002ae3f0 000007fe`f8a1c894 mscorwks!CTPMethodTable::OnCall+0x3dd7c3
				17 00000000`002ae6b0 000007ff`00191945 mscorwks!TransparentProxyStub_CrossContext+0x64
				18 00000000`002ae790 000007ff`001917f9 XXXXX_AgilePoint_WCFClient!.XXXXX.AgilePoint.WCFClient.WCFWorkflowProxy..ctor(System.String, System.String, System.String, System.String, System.Net.ICredentials, System.String)+0xf5

		- Exception

				Address: 0x00000000015f8b18
				HResult: 0x80131505
				Type: System.TimeoutException
				Message: The HTTP request to 'http://xxxxxx:9000/AgilePointServer/Workflow' has exceeded the allotted timeout of 00:00:00. The time allotted to this operation may have been a portion of a longer timeout.
				Stack Trace:
				SP               IP               Function                                                                                                                                                 Source
				00000000002ad130 000007fef5ca0c4a System.ServiceModel.Channels.HttpChannelUtilities.SetRequestTimeout(System.Net.HttpWebRequest, System.TimeSpan)                                          f:\dd\ndp\cdf\src\WCF\ServiceModel\System\ServiceModel\Channels\HttpChannelHelpers.cs @ 2484
				00000000002ad180 000007fef4deca91 System.ServiceModel.Channels.HttpChannelFactory+HttpRequestChannel+HttpChannelRequest.SendRequest(System.ServiceModel.Channels.Message, System.TimeSpan) f:\dd\ndp\cdf\src\WCF\ServiceModel\System\ServiceModel\Channels\HttpChannelFactory.cs @ 1004
				00000000002ad240 000007fef4dec774 System.ServiceModel.Channels.RequestChannel.Request(System.ServiceModel.Channels.Message, System.TimeSpan)  

				The request channel timed out attempting to send after 00:01:00

	- Conclusion
		- It seems our client didn’t get response from 'http://xxxx:9000/AgilePointServer/Workflow' in 1 minute. So a System.TimeoutException was raised
		- Our codes didn’t catch this exception, so this exception was an unhandled exception

				// XXXXX.AgilePoint.WCFClient.WCFWorkflowProxy
				public WCFWorkflowProxy(string appName, string machineName, string locale, string username, ICredentials credential, string workflowBindingName)
				{
				    …
				    try
				    {
				        wCFWorkflowServiceClient.CheckAuthenticated(this.m_wci);
				    }
				    finally
				    {
				        …
				    }
				}

		- UnhandledException happened; but our process still waiting for objects (didn’t crash). Typically this is going to be a debugging process. To disable this behavior, remove the AeDebug key: http://support.microsoft.com/kb/188296
NOTE: On x64 machines, this value will need to be modified in the following registry locations. Only the first location needs to be modified on x86 systems.

				HKEY_LOCAL_MACHINE\Software\Microsoft\Windows NT\CurrentVersion\AeDebug
				HKEY_LOCAL_MACHINE\Software\WOW6432Node\Microsoft\Windows NT\CurrentVersion\AeDebug

			You can delete the AeDebug key to disable debugging process, so that your application will crash rather than hang if any unhandled exception happened next time.

	- Suggestion
		- You can add some exception handle logic, to avoid application crash, so that it could exit graceful when met a TimoutExcepiton.
	- More
		- Demo

				using System;
				using System.Collections.Generic;
				using System.Linq;
				using System.Text;
				using System.Threading.Tasks;
				
				namespace ConsoleApplication1
				{
				    class Program
				    {
				        static void Main(string[] args)
				        {
				            try
				            {
				                throw new Exception("Mock Exception");
				            }
				            finally
				            {
				                Console.WriteLine("Run finally");
				                Console.ReadLine();
				            }
				        }
				    }
				}

		- dd